river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+    layout_manager: *zriver.LayoutManagerV1,
+
+    /// Initialize a connection
+    pub fn init(self: *LayoutManager, parent: *Self, client: *wl.Client, version: u32, id: u32) !void {
+        self.layout_manager = zriver.LayoutManagerV1.create(client, version, id) catch {
+            self.layout_manager.postNoMemory();
+            log.crit(.river_layout, "out of memory", .{});
+            return Error.BindFailed;
+        };
+        self.parent = parent;
+        self.layout_manager.setHandler(*LayoutManager, handleRequest, null, self);
+    }
+
+    pub fn deinit(self: *LayoutManager) void {
+        self.layout_manager.destroy();
+    }
+
+    fn handleRequest(layout_manager: *zriver.LayoutManagerV1, request: zriver.LayoutManagerV1.Request, self: *LayoutManager) void {
+        switch (request) {
+            .destroy => {
+                self.deinit();
+
+                // Remove this client from the global clients list
+                const node = @fieldParentPtr(std.TailQueue(LayoutManager).Node, "data", self);
+                self.parent.layout_managers.remove(node);
+                util.gpa.destroy(node);
+            },
+
+            .get_river_layout => |req| {
+                // Ignore if the output is inert
+                const wlr_output = wlr.Output.fromWlOutput(req.output) orelse return;
+                const output = @intToPtr(*Output, wlr_output.data);
+
+                log.debug(.layout_manager, "bind layout '{}' on output '{}'", .{ req.namespace, output.wlr_output.name });
+
+                const node = util.gpa.create(std.TailQueue(Layout).Node) catch {
+                    layout_manager.getClient().postNoMemory();
+                    log.crit(.river_layout, "out of memory", .{});
+                    return;
+                };
+                node.data.init(req.namespace, output, layout_manager.getClient(), 1, req.id) catch {
+                    layout_manager.getClient().postNoMemory();
+                    log.crit(.river_layout, "out of memory", .{});
+                    util.gpa.destroy(node);
+                    return;
+                };
+                output.layouts.append(node);
+            },
+        }
+    }
+};
+
+global: *wl.Global,
+server_destroy: wl.Listener(*wl.Server) = undefined,
+layout_managers: std.TailQueue(LayoutManager) = .{},
+
+pub fn init(self: *Self, server: *Server) !void {
+    self.* = .{
+        .global = try wl.Global.create(server.wl_server, zriver.LayoutManagerV1, 1, *Self, self, bind),
+    };
+
+    self.server_destroy.setNotify(handleServerDestroy);
+    server.wl_server.addDestroyListener(&self.server_destroy);
+}
+
+fn handleServerDestroy(listener: *wl.Listener(*wl.Server), wl_server: *wl.Server) void {
+    const self = @fieldParentPtr(Self, "server_destroy", listener);
+    self.global.destroy();
+
+    // Cleanup all connected clients
+    while (self.layout_managers.pop()) |node| {
+        node.data.deinit();
+        util.gpa.destroy(node);
+    }
```

Comment:
Don't need this once you create a destroy handler for the resource as libwayland will destroy all resources automatically when the display is terminated.